File: calcphis.cpp
#include <math.h>
#include <complex>
#include "param.h"
#include "procedures.h"
#include "hamiltonians.h"
#include "sc_tensor.h"
#include "results.h"
#include <iostream>
#include <fstream>

using namespace std;
using namespace N;
using namespace N1;
using namespace N2;
using namespace N3;

//---------------------------------------------------------------------------//
//---------------------------------------------------------------------------//

int main(){

	int i;

	param_dat p1;
	p1.read_in();
	p1.initialise();

	radical_dat r1;
	r1.read_in(p1);

	radical_dat r2;
	r2.read_in(p1);

	hamiltonian h1(r1);
	hamiltonian h2(r2);

	h1.build(p1, r1);
	h2.build(p1, r2);

	h1.noise(p1);
	h2.noise(p1);

	spin_corr sc1(p1, h1);
	spin_corr sc2(p1, h2);

	calc_dat c(p1);

	for (i=0; i<p1.nt; i++) {

		sc1.calc_sc(p1, h1);
		sc2.calc_sc(p1, h2);

		c.calc_ps(p1, sc1, sc2);
		c.integrate_sy(p1);

	}

	cout << c.phis[0] << "\t" << c.phis[1] << endl;

	c.write_results(p1);

	return 0;
};

//---------------------------------------------------------------------------//
//---------------------------------------------------------------------------//

File: fortran_src/calcphis.f90
        program s_yield
        use param
        use hamiltonians
        use procedures
        use sc_tensor
        implicit none

!------------------------------------------------------------------------------!
        type(hamiltonian) :: ha, hb
        type(spin_corr) :: sca, scb
        type(radical_dat) :: da, db
        type(param_dat) :: p_d
        type(calc_dat) :: c_d

        integer :: i, j, k, t
        integer :: cr

!------------------------------------------------------------------------------!

        !time code
        call system_clock(count_rate = cr)

        !read in parameters
        call p_d%read_in()                                     
        call da%read_in(p_d,1)                                      
        call db%read_in(p_d,2)                                      
        call p_d%initialise() 
        call c_d%pauli_matrices() 
        call c_d%get_time(p_d)
        call omp_set_num_threads(p_d%ncores)

!------------------------------------------------------------------------------!

        !build hamiltonians
        call ha%build(c_d, da)
        call hb%build(c_d, db)
        call ha%noise(p_d)
        call hb%noise(p_d)

        !initialise arrays
        call sca%create(ha, p_d)
        call scb%create(hb, p_d)
        call system_clock(p_d%start)
       
!------------------------------------------------------------------------------!

        !propagate over nt time steps
        do i = 1, p_d%nt
                call sca%calc_sc(ha, p_d)
                call scb%calc_sc(hb, p_d)

                call c_d%calc_ps(p_d, sca%r_ab, scb%r_ab)
                call c_d%integrate_sy(p_d)
                
                if (mod(i,p_d%nt/100) == 0) then
                  call system_clock(p_d%finish)
                  print *, i/(p_d%nt/100), '% done'
                  print *, 'Time', (p_d%finish - p_d%start)/dble(cr)
                  call system_clock(p_d%start)
                end if
                call flush()

        end do

!------------------------------------------------------------------------------!

        !write output
        call p_d%write_log_file(da, db)
        call c_d%rms_amp(p_d)
        
        end program s_yield
